File: ruby_challenges/src/hanoi.rb
class Hanoi

  def initialize(disks)
    @disks = disks
    @peg1  = []
    @peg2  = []
    @peg3  = []
    set_up(disks)
    puts "Playing Towers of Hanoi for #{@disks} disks:"
    show_pegs
  end

  def solve
    move(@disks, @peg1, @peg2, @peg3)
    puts "\nSolved."
  end

  def move(disks, peg1, peg2, peg3)
    if disks > 0
      move(disks - 1, peg1, peg3, peg2)
      peg3.push(peg1.pop)
      show_pegs
      move(disks - 1, peg2, peg1, peg3)
    end
  end

  def set_up(disks)
    disks.downto(1) do |disk|
      @peg1.push(disk)
    end
  end

  def show_pegs
    puts "\npeg 1: #@peg1"
    puts "peg 2: #{@peg2}"
    puts "peg 3: #{@peg3}"
  end

end

Hanoi.new(3).solve

File: java_challenges/Hanoi/src/main/java/Hanoi.java
import java.util.Stack;

class Hanoi {

    private static int Disks = 4;

    private Board board = new Board(Disks);

    Stack<Integer> peg_1 = new Stack<>();
    Stack<Integer> peg_2 = new Stack<>();
    Stack<Integer> peg_3 = new Stack<>();

    public static void main(String[] args) throws InterruptedException {
        new Hanoi().solve();
    }

    private void solve() throws InterruptedException {
        setUp();
        board.start();
        board.show(this);
        move(Disks, peg_1, peg_2, peg_3);
        board.done();
    }

    private void move(int n,
                      Stack<Integer> sourcePeg,
                      Stack<Integer> helperPeg,
                      Stack<Integer> targetPeg) throws InterruptedException {
        if (n > 0) {
            move(n - 1, sourcePeg, targetPeg, helperPeg);
            targetPeg.push(sourcePeg.pop());
            board.show(this);
            move(n - 1, helperPeg, sourcePeg, targetPeg);
        }
    }

    private void setUp() {
        for (int i = Disks; i > 0; i--) {
            peg_1.push(i);
        }
    }
}